# Multiple sequence alignment and phylogenetic tree Reconstruction

## Introduction

In order to generate a phylogeny tree using the single copy orthologs, we have to prepare unaligned ortholog files across multiple species.
I started with the single-copy orthologs generated by Compleasm for each species in ```gene_marker.fasta``` files. Each ```gene_marker.fasta``` contains all the single copy orthologs of each species. 

First, I renamed each file to include the species identifier, e.g., ```D_<species>_gene_marker.fasta```. Then, using an ```awk``` command, I parsed each renamed file to create separate .fas files, with each file representing one ortholog gene across all species. Finally, I used ```sed``` to clean up the headers in each ```.fas``` file for consistency. Each ```.fas``` file now contains a ```FASTA``` header that is simply a species name, followed by amino acid sequences.

The example of the cammand that I used to generated unaligned ortholog files for each ortholog for each species and fix the headers using the commands below: 

```
awk 'BEGIN{RS=">"; FS="\n"} {m=(split($1,b,"_")); fnme=b[1]".fas"; n=split(FILENAME,a,"/"); print ">" a[2] "\n" $2 >> fnme; close(fnme);}' ../*fasta
sed -i -e 's/.fasta//g' *fas
```

We are using 1 ortholog gene to show how to align sequences using Mafft and generate a phylogeny tree of Drosophila using Iqtree.

Since we are using only one ortholog gene, we can run the steps directly from the command line. Please run ```interact -A Introtogds -t 3:00:00``` to start an interactive session on a compute node, allowing you to work directly on the node rather than submitting a batch job.

If you runnning more than 1 genes, pleace submit each steps as slurm job, the example of the job scripts can be found in ```/projects/intro2gds/I2GDS2024/pep-seq/scripts/iris```

## 1. Aligning protein sequence using Mafft v7.505 

MAFFT (Multiple Alignment using Fast Fourier Transform) is a widely-used tool for aligning multiple sequences, whether they are nucleotide or protein sequences.

### 1.1 Copy unaligned alignments into your own directory

Create and move to a new directory in your home directory.
``` 
mkdir PEP_g1
cd PEP_g1
 ```

Copy the unaligned FASTA file into the PEP_g1 .

```
cp /projects/intro2gds/I2GDS2024/pep-seq/rawdata/iris/demo/129755at50557.fas .
```


### 1.2 Install [Mafft](https://mafft.cbrc.jp/alignment/software/) using Conda

```
module load Anaconda3 
conda create -n mafft -c conda-forge -c bioconda mafft
```
### 1.3 Activate the enviornment 

```conda activate mafft```

Use ```mafft --version``` to ensure the version that you are installing


### 1.4 Run Mafft 

```
mafft-linsi 129755at50557.fas > 129755at50557_aligned.fas
```

After running Mafft, you have a output file called ```129755at50557_aligned.fas``` in your unaligned_aa directory.

If you want to process more than one file at a time, you can use an sbatch array. First, we need to have a list of the files that we want to analyze. To do this, you can list the FASTA files in the ```unaligned_aa``` directory and re-direct them to a new text file. Also, you have to check how many lines there are in the locus_names.txt file. That is how many tasks you will need to feed your array. The array job script for running Mafft can be found in ```/projects/intro2gds/I2GDS2024/pep-seq/scripts/iris/arry_mafft.job```

```
ls *fas > locus_names.txt
wc -l locus_names.txt
```

## 2. Generate a phylogeny tree using IQTREE 2.3.6

Create a sub directory call ```iqtree``` in PEP_g1 and copy the aligned faa to that directory.

```
mkdir iqtree
cp 129755at50557_aligned.fas iqtree
cd iqtree
```

### 2.2 Install [IQTREE](http://www.iqtree.org/doc/)

Download the IQTREE 2.3.6
```
wget https://github.com/iqtree/iqtree2/releases/download/v2.3.6/iqtree-2.3.6-Linux-intel.tar.gz
tar -zxvf iqtree-2.3.6-Linux-intel.tar.gz
```
The software can also be founded in ```/projects/intro2gds/I2GDS2024/pep-seq/software/iris/iqtree-2.3.6-Linux-intel```

### 2.3 Choosing the right substitution model and recroustite the tree using
Starting with version 1.5.4 model selection option becomes a defult so we can simply run the following command  

```
iqtree-2.3.6-Linux-intel/bin/iqtree2 -s 129755at50557_aligned.fas
```
For 1 ortholog gene from 46 species, it take less than a minute to complete. Once it completed, you will have 8 output files, and we are focusing the flies listed below:

```129755at50557_aligned.fas.iqtree``` : Computational results, such as the staticaical data of the sequence alignment and result of choosing the best fit model.

```129755at50557_aligned.fas.treefile``` : The maximum-likelihood tree in NEWICK format.

For more information of the output files, you can go to [Tutorial#first-running-example](http://www.iqtree.org/doc/Tutorial#first-running-example)

The expected output files can be founded in ```/projects/intro2gds/I2GDS2024/pep-seq/result/iris/```
If you are running a large data set, please run it as a slurm job. job script for running iqtree2 can be found in ```/projects/intro2gds/I2GDS2024/pep-seq/scripts/iris/iqtree.job```

## Reference
Bui Quang Minh, Heiko A. Schmidt, Olga Chernomor, Dominik Schrempf, Michael D. Woodhams, Arndt von Haeseler, and Robert Lanfear (2020) IQ-TREE 2: New models and efficient methods for phylogenetic inference in the genomic era. Mol. Biol. Evol., in press. https://doi.org/10.1093/molbev/msaa015

Katoh, K., Misawa, K., Kuma, K. I., & Miyata, T. (2002). MAFFT: a novel method for rapid multiple sequence alignment based on fast Fourier transform. Nucleic acids research, 30(14), 3059-3066.




